File: kairospy/infrastructure/integrations/services/connections/connection.py
from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Protocol

from kairospy.infrastructure.integrations.application.connections import (
    IntegrationConnection,
    IntegrationConnectionSpec,
    ManagedResource,
)
from kairospy.infrastructure.integrations.domain import (
    AccessScope,
    ConnectionHealth,
    ConnectionIdentity,
    ConnectionLifecycle,
    ConnectionState,
    TransportKind,
)


class ConnectionComponent(ManagedResource, Protocol):
    """A lifecycle-managed resource owned by one physical link."""


@dataclass(slots=True)
class Connection(IntegrationConnection):
    """Base implementation for one transport link."""

    spec: IntegrationConnectionSpec
    components: tuple[ConnectionComponent, ...] = ()
    _state: ConnectionState = field(init=False)

    def __post_init__(self) -> None:
        identity = ConnectionIdentity(
            connection_id=self.spec.connection_id,
            route=self.spec.route,
            product=self.spec.product,
            asset_type=self.spec.asset_type,
            access=self.spec.access,
            transport=self.spec.transport,
            capability=self.spec.capability,
        )
        self._state = ConnectionState(identity=identity, bindings=self.spec.bindings)

    @property
    def identity(self) -> ConnectionIdentity:
        return self._state.identity

    @property
    def state(self) -> ConnectionState:
        return self._state

    @property
    def access(self) -> AccessScope:
        return self.spec.access

    @property
    def transport(self) -> TransportKind:
        return self.spec.transport

    async def start(self) -> None:
        if self._state.lifecycle is ConnectionLifecycle.READY:
            return
        self._state = _transition(self._state, ConnectionLifecycle.STARTING)
        started: list[ConnectionComponent] = []
        try:
            for component in self.components:
                await component.start()
                started.append(component)
        except Exception as error:
            for component in reversed(started):
                try:
                    await component.stop()
                except Exception:
                    pass
            self._state = _failed(self._state, error)
            raise
        self._state = _transition(
            self._state,
            ConnectionLifecycle.READY,
            authenticated=self.spec.access is AccessScope.PRIVATE and self.spec.credential is not None,
            connected_at=datetime.now(timezone.utc),
        )

    async def stop(self) -> None:
        if self._state.lifecycle in {ConnectionLifecycle.STOPPED, ConnectionLifecycle.CREATED}:
            self._state = _transition(self._state, ConnectionLifecycle.STOPPED)
            return
        self._state = _transition(self._state, ConnectionLifecycle.STOPPING)
        try:
            for component in reversed(self.components):
                await component.stop()
        except Exception as error:
            self._state = _failed(self._state, error)
            raise
        self._state = _transition(self._state, ConnectionLifecycle.STOPPED)

    async def reconnect(self) -> None:
        self._state = _transition(self._state, ConnectionLifecycle.STARTING)
        try:
            for component in self.components:
                await component.reconnect()
        except Exception as error:
            self._state = _failed(self._state, error)
            raise
        self._state = _transition(
            self._state,
            ConnectionLifecycle.READY,
            authenticated=self.spec.access is AccessScope.PRIVATE and self.spec.credential is not None,
            connected_at=datetime.now(timezone.utc),
            reconnect_count=self._state.reconnect_count + 1,
        )

    def health(self) -> ConnectionHealth:
        lifecycle = self._state.lifecycle
        return ConnectionHealth(
            lifecycle=lifecycle,
            healthy=lifecycle in {ConnectionLifecycle.READY, ConnectionLifecycle.DEGRADED},
            authenticated=self._state.authenticated,
            last_error=self._state.last_error,
        )


@dataclass(slots=True)
class CompositeConnectionComponent:
    name: str
    children: tuple[ConnectionComponent, ...] = ()

    async def start(self) -> None:
        for child in self.children:
            await child.start()

    async def stop(self) -> None:
        for child in reversed(self.children):
            await child.stop()

    async def reconnect(self) -> None:
        for child in self.children:
            await child.reconnect()


def _transition(state: ConnectionState, lifecycle: ConnectionLifecycle, **changes: object) -> ConnectionState:
    values = {
        "identity": state.identity,
        "lifecycle": lifecycle,
        "bindings": state.bindings,
        "authenticated": state.authenticated,
        "connected_at": state.connected_at,
        "last_error": state.last_error,
        "reconnect_count": state.reconnect_count,
    }
    values.update(changes)
    return ConnectionState(**values)


def _failed(state: ConnectionState, error: Exception) -> ConnectionState:
    return _transition(state, ConnectionLifecycle.FAILED, last_error=str(error))


__all__ = ["CompositeConnectionComponent", "ConnectionComponent", "Connection"]

File: crates/kairos-integration/src/services/connections/connection.rs
//! Lifecycle abstraction for one physical provider connection.
//!
//! A connection hides its HTTP/WS client, SDK, retry policy and subscriptions.
//! Business actors see only the capability-specific trait they receive from a
//! gateway; they never own provider transport details.

use crate::application::connection::Connection;
use crate::domain::{ConnectionHealth, ConnectionLifecycle, ConnectionSpec, ConnectionState};

pub trait ConnectionComponent: Send {
    fn start(&mut self) -> Result<(), String>;
    fn stop(&mut self) -> Result<(), String>;
    fn reconnect(&mut self) -> Result<(), String>;
}

pub struct ManagedConnection {
    spec: ConnectionSpec,
    state: ConnectionState,
    components: Vec<Box<dyn ConnectionComponent>>,
}

impl ManagedConnection {
    pub fn new(
        spec: ConnectionSpec,
        components: Vec<Box<dyn ConnectionComponent>>,
    ) -> Result<Self, String> {
        spec.validate()?;
        let identity = crate::domain::ConnectionIdentity::new(
            spec.connection_id.clone(),
            spec.provider.clone(),
            spec.product,
            spec.access,
            spec.transport,
            spec.capability,
        )?;
        Ok(Self {
            state: ConnectionState::new(identity),
            spec,
            components,
        })
    }

    fn fail(&mut self, error: String) -> Result<(), String> {
        self.state.lifecycle = ConnectionLifecycle::Failed;
        self.state.last_error = Some(error.clone());
        Err(error)
    }
}

impl Connection for ManagedConnection {
    fn identity(&self) -> &crate::domain::ConnectionIdentity {
        &self.state.identity
    }

    fn state(&self) -> &ConnectionState {
        &self.state
    }

    fn start(&mut self) -> Result<(), String> {
        if self.state.lifecycle == ConnectionLifecycle::Ready {
            return Ok(());
        }
        self.state.lifecycle = ConnectionLifecycle::Starting;
        let mut started = 0;
        for component in &mut self.components {
            if let Err(error) = component.start() {
                for component in self.components[..started].iter_mut().rev() {
                    let _ = component.stop();
                }
                return self.fail(error);
            }
            started += 1;
        }
        self.state.lifecycle = ConnectionLifecycle::Ready;
        self.state.authenticated = self.spec.credential_id.is_some();
        Ok(())
    }

    fn stop(&mut self) -> Result<(), String> {
        if matches!(
            self.state.lifecycle,
            ConnectionLifecycle::Created | ConnectionLifecycle::Stopped
        ) {
            self.state.lifecycle = ConnectionLifecycle::Stopped;
            return Ok(());
        }
        self.state.lifecycle = ConnectionLifecycle::Stopping;
        for component in self.components.iter_mut().rev() {
            if let Err(error) = component.stop() {
                return self.fail(error);
            }
        }
        self.state.lifecycle = ConnectionLifecycle::Stopped;
        Ok(())
    }

    fn reconnect(&mut self) -> Result<(), String> {
        self.state.lifecycle = ConnectionLifecycle::Starting;
        for component in &mut self.components {
            if let Err(error) = component.reconnect() {
                return self.fail(error);
            }
        }
        self.state.lifecycle = ConnectionLifecycle::Ready;
        self.state.reconnect_count += 1;
        self.state.last_error = None;
        Ok(())
    }

    fn health(&self) -> ConnectionHealth {
        ConnectionHealth {
            lifecycle: self.state.lifecycle,
            healthy: matches!(
                self.state.lifecycle,
                ConnectionLifecycle::Ready | ConnectionLifecycle::Degraded
            ),
            authenticated: self.state.authenticated,
            last_error: self.state.last_error.clone(),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::{Connection, ConnectionComponent, ManagedConnection};
    use crate::application::ConnectionSpec;
    use crate::domain::{AccessScope, IntegrationCapability, ProductFamily, TransportKind};

    struct Component {
        starts: usize,
        stops: usize,
    }

    impl ConnectionComponent for Component {
        fn start(&mut self) -> Result<(), String> {
            self.starts += 1;
            Ok(())
        }
        fn stop(&mut self) -> Result<(), String> {
            self.stops += 1;
            Ok(())
        }
        fn reconnect(&mut self) -> Result<(), String> {
            self.starts += 1;
            Ok(())
        }
    }

    #[test]
    fn managed_connection_owns_component_lifecycle() {
        let spec = ConnectionSpec {
            connection_id: "reference.binance.spot.rest".into(),
            provider: "binance".into(),
            product: Some(ProductFamily::Spot),
            access: AccessScope::Public,
            transport: TransportKind::Rest,
            capability: IntegrationCapability::Reference,
            credential_id: None,
            asset_type: None,
        };
        let mut connection = ManagedConnection::new(
            spec,
            vec![Box::new(Component {
                starts: 0,
                stops: 0,
            })],
        )
        .unwrap();
        assert_eq!(
            connection.state().lifecycle,
            crate::domain::ConnectionLifecycle::Created
        );
        connection.start().unwrap();
        assert!(connection.health().healthy);
        connection.reconnect().unwrap();
        assert_eq!(connection.state().reconnect_count, 1);
        connection.stop().unwrap();
        assert_eq!(
            connection.state().lifecycle,
            crate::domain::ConnectionLifecycle::Stopped
        );
    }
}
